### PR TITLE
feat(medusa): Improve config loading

### PR DESCRIPTION
**Pull request by adrien2p**
At the moment, the way the config and the env are loaded only allow to target one `.env` file.
In many cases, as a user, you want to be able to run your server with either, local, prod-like and tests env.
Also, the env was loaded as a side effect at a specific time which is not necessarily the moment you need them (passport/express loader).

That PR loads the env immediately in the loader index file using the config side effect.
It allows to work wit `.env`, `.env.local` and `.env.test`. Anyway, for backward compatibility, there is always a fallback on the `.env` file if the one related to the environment is not present.

This PR also clean and remove env that was legacy.

### CODE REVIEW

**Review by srindom — COMMENTED** [review]
Would it make sense to move everything related to config to be configurable by `medusa-config.js` and then have `loaders/config` pull in values from there with defaults in case values are not provided. This way we remove `process.env` usage in the core and move it to the project level. 

We will also eliminate everything opinionated about what your .env files should be called as you would now use:

```js
// medusa-config.js

dotenv.config([whatever your file is])

exports.module = {
 ... usage of your .env how ever you want 
}
```

**Comment by adrien2p** [comment]
> Would it make sense to move everything related to config to be configurable by `medusa-config.js` and then have `loaders/config` pull in values from there with defaults in case values are not provided. This way we remove `process.env` usage in the core and move it to the project level.
> 
> We will also eliminate everything opinionated about what your .env files should be called as you would now use:
> 
> ```js
> // medusa-config.js
> 
> dotenv.config([whatever your file is])
> 
> exports.module = {
>  ... usage of your .env how ever you want 
> }
> ```

I am agree, should we remove the dotenv package from the core then?

**Comment by adrien2p** [comment]
@srindom is that alright even if it breaks the existing ? 

At the moment the users does not load there envs themselves

**Comment by adrien2p** [comment]
@srindom I am waiting to re run the tests because locally they are green 
Screenshot 2022-04-05 at 15 13 27

**Comment by adrien2p** [comment]
Fixes #827

File: packages/medusa/src/services/invite.ts [hunk 1]
```diff
@@ -62,22 +69,27 @@ class InviteService extends BaseService {
...
     }
 
-    const cloned = new InviteService({
-      manager,
-      inviteRepository: this.inviteRepository_,
-      userService: this.userService_,
-      userRepository: this.userRepo_,
-      eventBusService: this.eventBus_,
-    })
+    console.log("-------", JSON.stringify(this.configModule_, null, 4))
```
Comment (srindom) [inline packages/medusa/src/services/invite.ts:72]: issue: rm

File: packages/medusa/src/services/invite.ts [hunk 2]
```diff
@@ -248,8 +260,10 @@ class InviteService extends BaseService {
   }
 
   verifyToken(token): JwtPayload | string {
-    if (config.jwtSecret) {
-      return jwt.verify(token, config.jwtSecret)
+    console.log("----", this.configModule_)
```
Comment (srindom) [inline packages/medusa/src/services/invite.ts:263]: issue: rm

File: packages/medusa/src/types/global.ts [hunk 3]
```diff
@@ -1,10 +1,39 @@
...
+export type Logger = _Logger & {
+  progress: (activityId: string, msg: string) => void
+}
+
+export type ConfigModule = {
+  projectConfig: {
+    redis_url?: string
+
+    jwtSecret?: string
+    cookieSecret?: string
```
Comment (srindom) [inline packages/medusa/src/types/global.ts:22]: todo: snake_case

File: packages/medusa/src/types/global.ts [hunk 4]
```diff
@@ -1,10 +1,39 @@
...
+    database_extra?: Record<string, unknown> & {
+      ssl: { rejectUnauthorized: false }
+    }
+    store_cors?: string
+    admin_cors?: string
+  }
+  plugins: {
+    resolve: string
+    options: Record<string, unknown>
+  }[]
```
Comment (srindom) [inline packages/medusa/src/types/global.ts:38]:
```suggestion
  plugins: ({
    resolve: string
    options: Record<string, unknown>
  } | string)[]
```

**Review by srindom — CHANGES_REQUESTED** [review]

File: packages/medusa/src/loaders/config.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+const errorHandler = isProduction ? (msg: string) => {
+  throw new Error(msg)
+} : console.log
+
+export default (rootDirectory: string): ConfigModule => {
+  const { configModule } = getConfigFile(rootDirectory, `medusa-config`) as {
+    configModule: ConfigModule
+  }
+
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/loaders/config.ts:14]: do you mean something like that?

```
return {
    projectConfig: {
      jwt_secret:
        process.env.JWT_SECRET ?? process.env.COOKIE_SECRET ?? "supersecret",
      cookie_secret:
        process.env.JWT_SECRET ?? process.env.COOKIE_SECRET ?? "supersecret",
      ...configModule?.projectConfig,
    },
    plugins: configModule?.plugins ?? [],
  }
}
```

File: packages/medusa/src/loaders/config.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  if (!configModule?.projectConfig?.redis_url) {
+    console.log(
+      `[medusa-config] ⚠️ redis_url not found. A fake redis instance will be used.`
+    )
+  }
+
+  if (!configModule?.projectConfig?.jwtSecret) {
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/loaders/config.ts:21]: should not happen with the remarks above if I set it directly from the loader

**Comment by adrien2p** [comment]
@srindom this branch is up to date with develop but still fail with the same tests as some other PR

**Review by srindom — COMMENTED** [review]
LGTM - will let Oli approve too :)
cc: @olivermrbl

**Review by srindom — APPROVED** [review]

**Review by olivermrbl — APPROVED** [review]
🙌

**Comment by olivermrbl** [comment]
Ready to merge this one @adrien2p ?

**Comment by adrien2p** [comment]
Yes 🙌🚀 when you want

**Comment by olivermrbl** [comment]
Confirming with @srindom?

**Pull request merged**